File: src/views/MetadataPanel.m
/**
 * MetadataPanel.m
 * Metadata panel implementation
 */

#import "MetadataPanel.h"
#import "../models/PhotoStore.h"
#import "../models/PhotoItem.h"

/* Flipped view for top-to-bottom layout */
@interface FlippedView : NSView
@end

@implementation FlippedView
- (BOOL)isFlipped {
    return YES;
}
@end

@interface MetadataPanel ()

@property (nonatomic, weak, readwrite) PhotoStore *photoStore;
@property (nonatomic, strong) NSScrollView *scrollView;
@property (nonatomic, strong) NSStackView *contentStack;
@property (nonatomic, strong) FlippedView *documentView;

@end

@implementation MetadataPanel

- (instancetype)initWithPhotoStore:(PhotoStore *)photoStore {
    self = [super initWithFrame:NSZeroRect];
    if (self) {
        _photoStore = photoStore;
        [self setupViews];
    }
    return self;
}

- (void)setupViews {
    self.wantsLayer = YES;
    self.layer.backgroundColor = [NSColor colorWithRed:0.075 green:0.075 blue:0.085 alpha:1.0].CGColor;
    
    /* Scroll view */
    self.scrollView = [[NSScrollView alloc] initWithFrame:self.bounds];
    self.scrollView.translatesAutoresizingMaskIntoConstraints = NO;
    self.scrollView.hasVerticalScroller = YES;
    self.scrollView.autohidesScrollers = YES;
    self.scrollView.borderType = NSNoBorder;
    self.scrollView.backgroundColor = [NSColor colorWithRed:0.075 green:0.075 blue:0.085 alpha:1.0];
    self.scrollView.drawsBackground = YES;
    [self addSubview:self.scrollView];
    
    /* Flipped document view for top alignment */
    self.documentView = [[FlippedView alloc] init];
    self.documentView.translatesAutoresizingMaskIntoConstraints = NO;
    self.scrollView.documentView = self.documentView;
    
    /* Content stack */
    self.contentStack = [[NSStackView alloc] init];
    self.contentStack.translatesAutoresizingMaskIntoConstraints = NO;
    self.contentStack.orientation = NSUserInterfaceLayoutOrientationVertical;
    self.contentStack.alignment = NSLayoutAttributeLeading;
    self.contentStack.spacing = 12;
    [self.documentView addSubview:self.contentStack];
    
    [NSLayoutConstraint activateConstraints:@[
        [self.scrollView.topAnchor constraintEqualToAnchor:self.topAnchor],
        [self.scrollView.leadingAnchor constraintEqualToAnchor:self.leadingAnchor],
        [self.scrollView.trailingAnchor constraintEqualToAnchor:self.trailingAnchor],
        [self.scrollView.bottomAnchor constraintEqualToAnchor:self.bottomAnchor],
        
        [self.contentStack.topAnchor constraintEqualToAnchor:self.documentView.topAnchor constant:16],
        [self.contentStack.leadingAnchor constraintEqualToAnchor:self.documentView.leadingAnchor constant:16],
        [self.contentStack.trailingAnchor constraintEqualToAnchor:self.documentView.trailingAnchor constant:-16],
        [self.contentStack.bottomAnchor constraintLessThanOrEqualToAnchor:self.documentView.bottomAnchor constant:-16],
        
        /* Document view width matches scroll view */
        [self.documentView.widthAnchor constraintEqualToAnchor:self.scrollView.widthAnchor],
    ]];
}

- (void)refresh {
    /* Clear existing content */
    for (NSView *view in self.contentStack.arrangedSubviews.copy) {
        [self.contentStack removeArrangedSubview:view];
        [view removeFromSuperview];
    }
    
    PhotoItem *photo = [self.photoStore currentPhoto];
    if (!photo) {
        return;
    }
    
    /* Load metadata if needed */
    [photo loadMetadata];
    
    /* File name header */
    NSTextField *nameLabel = [NSTextField labelWithString:photo.name];
    nameLabel.font = [NSFont systemFontOfSize:13 weight:NSFontWeightMedium];
    nameLabel.textColor = [NSColor colorWithWhite:0.9 alpha:1.0];
    nameLabel.lineBreakMode = NSLineBreakByTruncatingMiddle;
    [self.contentStack addArrangedSubview:nameLabel];
    
    [self addSeparator];
    
    /* File info */
    [self addSectionTitle:@"FILE"];
    [self addRow:@"Size" value:[PhotoItem formattedFileSize:photo.fileSize]];
    [self addRow:@"Created" value:[self formatDate:photo.createdDate]];
    
    /* Image info */
    [self addSeparator];
    [self addSectionTitle:@"IMAGE"];
    if (photo.dimensions.width > 0) {
        NSString *dims = [NSString stringWithFormat:@"%.0f x %.0f", 
                         photo.dimensions.width, photo.dimensions.height];
        [self addRow:@"Dimensions" value:dims];
    }
    if (photo.colorSpace) {
        [self addRow:@"Color Space" value:photo.colorSpace];
    }
    if (photo.dateTaken) {
        [self addRow:@"Date Taken" value:[self formatDate:photo.dateTaken]];
    }
    
    /* Camera info */
    if (photo.cameraMake || photo.cameraModel) {
        [self addSeparator];
        [self addSectionTitle:@"CAMERA"];
        if (photo.cameraMake) {
            [self addRow:@"Make" value:photo.cameraMake];
        }
        if (photo.cameraModel) {
            [self addRow:@"Model" value:photo.cameraModel];
        }
    }
    
    /* Exposure info */
    if (photo.exposureTime || photo.fNumber || photo.iso || photo.focalLength) {
        [self addSeparator];
        [self addSectionTitle:@"EXPOSURE"];
        if (photo.exposureTime) {
            [self addRow:@"Shutter" value:photo.exposureTime];
        }
        if (photo.fNumber) {
            [self addRow:@"Aperture" value:photo.fNumber];
        }
        if (photo.iso) {
            [self addRow:@"ISO" value:[NSString stringWithFormat:@"%@", photo.iso]];
        }
        if (photo.focalLength) {
            [self addRow:@"Focal Length" value:photo.focalLength];
        }
    }
    
    /* Force layout update */
    [self.documentView setNeedsLayout:YES];
    [self.documentView layoutSubtreeIfNeeded];
    
    /* Update document view height to fit content */
    CGFloat contentHeight = self.contentStack.fittingSize.height + 32;
    [self.documentView setFrameSize:NSMakeSize(self.scrollView.contentSize.width, contentHeight)];
}

#pragma mark - Helpers

- (void)addSectionTitle:(NSString *)title {
    NSTextField *label = [NSTextField labelWithString:title];
    label.font = [NSFont systemFontOfSize:10 weight:NSFontWeightSemibold];
    label.textColor = [NSColor colorWithWhite:0.45 alpha:1.0];
    [self.contentStack addArrangedSubview:label];
}

- (void)addRow:(NSString *)labelText value:(NSString *)value {
    if (!value) {
        return;
    }
    
    NSStackView *row = [[NSStackView alloc] init];
    row.orientation = NSUserInterfaceLayoutOrientationHorizontal;
    row.distribution = NSStackViewDistributionFill;
    row.spacing = 8;
    
    NSTextField *label = [NSTextField labelWithString:labelText];
    label.font = [NSFont systemFontOfSize:12 weight:NSFontWeightRegular];
    label.textColor = [NSColor colorWithWhite:0.5 alpha:1.0];
    [label setContentHuggingPriority:NSLayoutPriorityDefaultHigh 
                      forOrientation:NSLayoutConstraintOrientationHorizontal];
    
    NSTextField *valueField = [NSTextField labelWithString:value];
    valueField.font = [NSFont monospacedSystemFontOfSize:11 weight:NSFontWeightRegular];
    valueField.textColor = [NSColor colorWithWhite:0.85 alpha:1.0];
    valueField.alignment = NSTextAlignmentRight;
    valueField.lineBreakMode = NSLineBreakByTruncatingMiddle;
    [valueField setContentCompressionResistancePriority:NSLayoutPriorityDefaultLow 
                                         forOrientation:NSLayoutConstraintOrientationHorizontal];
    
    [row addArrangedSubview:label];
    [row addArrangedSubview:valueField];
    
    [self.contentStack addArrangedSubview:row];
}

- (void)addSeparator {
    NSBox *separator = [[NSBox alloc] init];
    separator.boxType = NSBoxSeparator;
    separator.translatesAutoresizingMaskIntoConstraints = NO;
    [self.contentStack addArrangedSubview:separator];
}

- (NSString *)formatDate:(NSDate *)date {
    if (!date) {
        return nil;
    }
    
    static NSDateFormatter *formatter = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        formatter = [[NSDateFormatter alloc] init];
        formatter.dateStyle = NSDateFormatterMediumStyle;
        formatter.timeStyle = NSDateFormatterShortStyle;
    });
    
    return [formatter stringFromDate:date];
}

@end

File: src/swift/Views/MetadataPanel.swift
/**
 * MetadataPanel.swift
 * Glass sidebar showing photo metadata and EXIF info
 */

import SwiftUI

struct MetadataPanel: View {
    @Environment(PhotoStore.self) private var photoStore
    @State private var photo: PhotoItem?
    
    var body: some View {
        ScrollView {
            VStack(alignment: .leading, spacing: 0) {
                if let photo = photo {
                    metadataContent(for: photo)
                } else {
                    emptyState
                }
            }
            .padding(20)
        }
        .background {
            // Glass panel background
            Rectangle()
                .fill(.ultraThinMaterial.opacity(0.5))
                .overlay {
                    Rectangle()
                        .fill(
                            LinearGradient(
                                colors: [.white.opacity(0.08), .clear],
                                startPoint: .leading,
                                endPoint: .trailing
                            )
                        )
                }
        }
        .overlay(alignment: .leading) {
            // Left edge highlight
            Rectangle()
                .fill(.white.opacity(0.1))
                .frame(width: 1)
        }
        .onChange(of: photoStore.selectedIndex) { _, _ in
            loadCurrentPhoto()
        }
        .onChange(of: photoStore.photos) { _, _ in
            loadCurrentPhoto()
        }
        .onAppear {
            loadCurrentPhoto()
        }
    }
    
    private var emptyState: some View {
        VStack(spacing: 12) {
            Image(systemName: "info.circle")
                .font(.system(size: 32, weight: .light))
                .foregroundStyle(.white.opacity(0.2))
            
            Text("No photo selected")
                .font(.system(size: 13, weight: .medium))
                .foregroundStyle(.white.opacity(0.3))
        }
        .frame(maxWidth: .infinity, maxHeight: .infinity)
        .padding(.vertical, 60)
    }
    
    private func metadataContent(for photo: PhotoItem) -> some View {
        VStack(alignment: .leading, spacing: 24) {
            // File name header
            VStack(alignment: .leading, spacing: 8) {
                Text(photo.name)
                    .font(.system(size: 14, weight: .semibold))
                    .foregroundStyle(.white)
                    .lineLimit(2)
                
                Text(photo.path)
                    .font(.system(size: 11, weight: .medium, design: .monospaced))
                    .foregroundStyle(.white.opacity(0.3))
                    .lineLimit(1)
                    .truncationMode(.middle)
            }
            
            GlassDivider()
            
            // File info
            MetadataSection(title: "FILE") {
                MetadataRow(label: "Size", value: PhotoItem.formattedFileSize(photo.fileSize))
                MetadataRow(label: "Created", value: DateFormatters.formatForDisplay(photo.createdDate))
                MetadataRow(label: "Modified", value: DateFormatters.formatForDisplay(photo.modifiedDate))
            }
            
            // Image info
            if photo.dimensions.width > 0 {
                GlassDivider()
                
                MetadataSection(title: "IMAGE") {
                    MetadataRow(
                        label: "Dimensions",
                        value: "\(Int(photo.dimensions.width)) x \(Int(photo.dimensions.height))"
                    )
                    if let colorSpace = photo.colorSpace {
                        MetadataRow(label: "Color Space", value: colorSpace)
                    }
                    if let dateTaken = photo.dateTaken {
                        MetadataRow(label: "Date Taken", value: DateFormatters.formatForDisplay(dateTaken))
                    }
                }
            }
            
            // Camera info
            if photo.cameraMake != nil || photo.cameraModel != nil {
                GlassDivider()
                
                MetadataSection(title: "CAMERA") {
                    if let make = photo.cameraMake {
                        MetadataRow(label: "Make", value: make)
                    }
                    if let model = photo.cameraModel {
                        MetadataRow(label: "Model", value: model)
                    }
                }
            }
            
            // Exposure info
            if photo.exposureTime != nil || photo.fNumber != nil || photo.iso != nil || photo.focalLength != nil {
                GlassDivider()
                
                MetadataSection(title: "EXPOSURE") {
                    if let exposure = photo.exposureTime {
                        MetadataRow(label: "Shutter", value: exposure)
                    }
                    if let aperture = photo.fNumber {
                        MetadataRow(label: "Aperture", value: aperture)
                    }
                    if let iso = photo.iso {
                        MetadataRow(label: "ISO", value: "\(iso)")
                    }
                    if let focal = photo.focalLength {
                        MetadataRow(label: "Focal Length", value: focal)
                    }
                }
            }
            
            Spacer()
        }
    }
    
    private func loadCurrentPhoto() {
        if var currentPhoto = photoStore.currentPhoto {
            currentPhoto.loadMetadata()
            photo = currentPhoto
        } else {
            photo = nil
        }
    }
}

struct MetadataSection<Content: View>: View {
    let title: String
    @ViewBuilder let content: Content
    
    var body: some View {
        VStack(alignment: .leading, spacing: 12) {
            Text(title)
                .font(.system(size: 10, weight: .bold))
                .foregroundStyle(.white.opacity(0.4))
                .tracking(1.5)
            
            content
        }
    }
}

struct MetadataRow: View {
    let label: String
    let value: String
    
    var body: some View {
        HStack(alignment: .top) {
            Text(label)
                .font(.system(size: 12, weight: .medium))
                .foregroundStyle(.white.opacity(0.5))
            
            Spacer()
            
            Text(value)
                .font(.system(size: 12, weight: .medium, design: .monospaced))
                .foregroundStyle(.white.opacity(0.8))
                .multilineTextAlignment(.trailing)
        }
    }
}

struct GlassDivider: View {
    var body: some View {
        Rectangle()
            .fill(
                LinearGradient(
                    colors: [.clear, .white.opacity(0.15), .clear],
                    startPoint: .leading,
                    endPoint: .trailing
                )
            )
            .frame(height: 1)
    }
}
